### PR TITLE
Improve --prof-exec infrastructure and report

### PR DESCRIPTION
**Pull request by gezalore**
Again --prof-exec have bit-rotted a little with all the recent changes to the structure of the generated code. This patch contains a few improvements:
- Repalce the eval/evl_loop begin/end events with generic section_push/section_pop events, that can be arbitrarily sprinkled into the generate code (so long as they are matched correctly) to measure various sections. The report then contains a nested profile of the sections, and the VCD trace shows the section names.
- Better handling of exec graphs
- Clearer overall statistics

### CODE REVIEW

**Comment by gezalore** [comment]
Here is an example trace and report (t_gantt.pl --vltmt):
image

```
Verilator Gantt report

Argument settings:
  +verilator+prof+exec+start+2
  +verilator+prof+exec+window+2

Summary:
  Total elapsed time = 19856 rdtsc ticks
  Parallelized code  = 42.41% of elapsed time
  Total threads      = 2
  Total CPUs used    = 2
  Total mtasks       = 7
  Total yields       = 0

Parallelized code, measured:
  Thread utilization =  22.07%
  Speedup            =  0.441x

Parallelized code, predicted during static scheduling:
  Thread utilization =  62.56%
  Speedup            =   1.25x

All code, measured:
  Thread utilization =  38.15%
  Speedup            =  0.763x

All code, measured, scaled by predicted speedup:
  Thread utilization =  52.59%
  Speedup            =   1.05x

MTask statistics:
  Longest mtask id = 2
  Longest mtask time = 11.54% of time elapsed in parallelized code
  min log(p2e) = -3.621  from mtask 2 (predict 26, elapsed 972)
  max log(p2e) = -2.096  from mtask 5 (predict 103, elapsed 838)
  mean = -2.711
  stddev = 0.467
  e ^ stddev = 1.595

CPU info:
   Id | MTask   | Socket | Core | Model
      | time    |        |      |
  ====|=========|========|======|======
    1 |  14.09% |      0 |    1 | 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
    4 |   4.62% |      0 |    4 | 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz

Section profile:
 Total    | Self    | Total    | Relative   | Section
 time     | time    | entries  | entries    |  name
==========|=========|==========|============|========
  100.00% |  16.90% |        1 |       1.00 | *TOTAL*
   83.10% |   7.22% |        2 |       2.00 |   eval
   75.88% |   9.69% |        2 |       1.00 |     loop nba
   45.35% |  45.35% |        1 |       0.50 |       func nba
   20.84% |  12.45% |        3 |       1.50 |       loop act
    7.33% |   7.33% |        4 |       1.33 |         trig act
    1.06% |   1.06% |        1 |       0.33 |         func act
```

File: src/V3EmitCModel.cpp [hunk 1]
```diff
@@ -422,7 +422,7 @@ class EmitCModel final : public EmitCFunc {
         }
         puts("Verilated::endOfEval(vlSymsp->__Vm_evalMsgQp);\n");
 
-        if (v3Global.opt.profExec()) puts("VL_EXEC_TRACE_ADD_RECORD(vlSymsp).evalEnd();\n");
+        if (v3Global.opt.profExec()) puts("VL_EXEC_TRACE_ADD_RECORD(vlSymsp).sectionPop();\n");
```
Comment (wsnyder) [inline src/V3EmitCModel.cpp:425]: Maybe have sectionPush be wrapped in a class which calls it, then when destroyed automatically calls sectionPop?  (Likewise cleans up code elsewhere).

File: test_regress/t/t_gantt_io.out [hunk 2]
```diff
@@ -4,33 +4,44 @@ Argument settings:
...
 
-CPUs:
-  cpu 10: cpu_time=4725 socket=0 core=10  Test Ryzen 9 3950X 16-Core Processor
-  cpu 19: cpu_time=770 socket=0 core=3  Test Ryzen 9 3950X 16-Core Processor
+CPU info:
+   Id | MTask   | Socket | Core | Model
+      | time    |        |      |
+  ====|=========|========|======|======
+   10 |  20.18% |      0 |   10 | Test Ryzen 9 3950X 16-Core Processor
+   19 |   3.29% |      0 |    3 | Test Ryzen 9 3950X 16-Core Processor
```
Comment (wsnyder) [inline test_regress/t/t_gantt_io.out:45]: Can you show the raw cpu time too please?

**Review by wsnyder — COMMENTED** [review]
Appreciated!

Please manually run "make lint-py" if you didn't already.

File: test_regress/t/t_gantt_io.out (hunk 2 above)
Comment (gezalore) [reply test_regress/t/t_gantt_io.out:45]: Does this work for you?
```
CPU info:
   Id | Time spent executing MTask | Socket | Core | Model
      | % of elapsed ticks / ticks |        |      |
  ====|============================|========|======|======
    4 |   1.80% /              278 |      0 |    4 | 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
   14 |  27.74% /             4288 |      0 |    6 | 11th Gen Intel(R) Core(TM) i9-11950H @ 2.60GHz
```

File: src/V3EmitCModel.cpp (hunk 1 above)
Comment (gezalore) [reply src/V3EmitCModel.cpp:425]: I agree that would be nice, but it does not work nicely for all use cases. Currently `eval` have multiple push/pop, with some pop not strictly on scope exit:

```
   void Vt_gantt___024root___eval(Vt_gantt___024root* vlSelf) {
       if (false && vlSelf) {}  // Prevent unused
       Vt_gantt__Syms* const __restrict vlSymsp VL_ATTR_UNUSED = vlSelf->vlSymsp;
       VL_DEBUG_IF(VL_DBG_MSGF("+    Vt_gantt___024root___eval\n"); );
       // Init
       VlTriggerVec<1> __VpreTriggered;
       IData/*31:0*/ __VnbaIterCount;
       CData/*0:0*/ __VnbaContinue;
       // Body
       VL_EXEC_TRACE_ADD_RECORD(vlSymsp).sectionPush("loop nba");
       __VnbaIterCount = 0U;
       __VnbaContinue = 1U;
       while (__VnbaContinue) {
           __VnbaContinue = 0U;
           vlSelf->__VnbaTriggered.clear();
           VL_EXEC_TRACE_ADD_RECORD(vlSymsp).sectionPush("loop act");
           vlSelf->__VactIterCount = 0U;
           vlSelf->__VactContinue = 1U;
           while (vlSelf->__VactContinue) {
               vlSelf->__VactContinue = 0U;
               Vt_gantt___024root___eval_triggers__act(vlSelf);
               if (vlSelf->__VactTriggered.any()) {
                   vlSelf->__VactContinue = 1U;
                   if (VL_UNLIKELY((0x64U < vlSelf->__VactIterCount))) {
   #ifdef VL_DEBUG
                       Vt_gantt___024root___dump_triggers__act(vlSelf);
   #endif
                       VL_FATAL_MT("t/t_gen_alw.v", 6, "", "Active region did not converge.");
                   }
                   vlSelf->__VactIterCount = ((IData)(1U)`
                                              + vlSelf->__VactIterCount);
                   __VpreTriggered.andNot(vlSelf->__VactTriggered, vlSelf->__VnbaTriggered);
                   vlSelf->__VnbaTriggered.thisOr(vlSelf->__VactTriggered);
                   Vt_gantt___024root___eval_act(vlSelf);
               }
           }
           VL_EXEC_TRACE_ADD_RECORD(vlSymsp).sectionPop();
           if (vlSelf->__VnbaTriggered.any()) {
               __VnbaContinue = 1U;
               if (VL_UNLIKELY((0x64U < __VnbaIterCount))) {
   #ifdef VL_DEBUG
                   Vt_gantt___024root___dump_triggers__nba(vlSelf);
   #endif
                   VL_FATAL_MT("t/t_gen_alw.v", 6, "", "NBA region did not converge.");
               }
               __VnbaIterCount = ((IData)(1U) + __VnbaIterCount);
               Vt_gantt___024root___eval_nba(vlSelf);
           }
       }
       VL_EXEC_TRACE_ADD_RECORD(vlSymsp).sectionPop();
   }
```

File: test_regress/t/t_gantt_io.out (hunk 2 above)
Comment (wsnyder) [reply test_regress/t/t_gantt_io.out:45]: Yes thanks

**Comment by wsnyder** [comment]
BTW I'll shortly push cleaning up some other pylint warnings which show up in other files when did a pip upgrade to the latest.

**Pull request merged**